### PR TITLE
Support logging PyTorch metrics on training batch steps with autolog

### LINKED ISSUE #4235: [BUG] mlflow.autolog() for PyTorch Lightning logs `_step` metrics for epochs instead of steps
### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: Reproducible with example from [official docs](https://www.mlflow.org/docs/latest/python_api/mlflow.pytorch.html)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Docker image: `nvidia/cuda:10.2-base-ubuntu16.04`
- **MLflow installed from (source or binary)**: Python package
- **MLflow version (run ``mlflow --version``)**: 1.15.0
- **Python version**: 3.8.8
- **npm version, if running the dev UI**: No
- **Exact command to reproduce**: See code snippet

### Describe the problem
Expected behavior:
If I'm using `mlflow.autolog()` with a `pl.LightningModule` and use `self.log("train_loss", loss, on_epoch=True, on_step=True, logger=True)` inside its `training_step` function, MLflow should log `train_loss_step` metric for every logged **step** and `train_loss_epoch` for every logged **epoch**

Actual behavior:
`train_loss_step` and `train_loss_epoch` are recorded for every logged **epoch** only.

After running code in below sample, `train_loss_epoch`  contains 1 entry per epoch:
```
1617996945494 1.3506861925125122 0
1617996966448 1.201533317565918 1
1617996986735 1.1901590824127197 2
1617997007487 1.190608024597168 3
1617997028088 1.1838353872299194 4
1617997048012 1.1807078123092651 5
1617997068702 1.18239426612854 6
1617997089547 1.1766492128372192 7
1617997110542 1.175412654876709 8
1617997130747 1.1763886213302612 9
1617997150977 1.1739486455917358 10
1617997171487 1.1722205877304077 11
1617997191321 1.170091152191162 12
1617997211502 1.1701279878616333 13
1617997232425 1.1694077253341675 14
1617997252996 1.1664137840270996 15
1617997273641 1.1662592887878418 16
1617997294056 1.1691796779632568 17
1617997314113 1.164090871810913 18
1617997335002 1.1653193235397339 19
```
`train_loss_step` also contains 1 entry per **epoch**
```
1617996945494 1.0488566160202026 0
1617996966448 1.0117261409759521 1
1617996986735 0.9437647461891174 2
1617997007487 1.1345815658569336 3
1617997028088 0.9712427854537964 4
1617997048012 0.9615302681922913 5
1617997068702 0.9383407235145569 6
1617997089547 0.9366243481636047 7
1617997110542 0.9363885521888733 8
1617997130747 0.9403294324874878 9
1617997150977 0.9360908269882202 10
1617997171487 1.0088082551956177 11
1617997191321 0.9398048520088196 12
1617997211502 0.9855148792266846 13
1617997232425 0.9356205463409424 14
1617997252996 0.9355267286300659 15
1617997273641 0.937127947807312 16
1617997294056 0.9390522837638855 17
1617997314113 0.9355039000511169 18
1617997335002 0.9383139610290527 19
```

### Code to reproduce issue
Small modifications to [code snippet in documentation](https://www.mlflow.org/docs/latest/python_api/mlflow.pytorch.html)

```
import os

import pytorch_lightning as pl
import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST

import mlflow.pytorch


class MNISTModel(pl.LightningModule):
    def __init__(self):
        super(MNISTModel, self).__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_nb):
        x, y = batch
        loss = F.cross_entropy(self(x), y)

        # Use the current of PyTorch logger
        self.log("train_loss", loss, on_epoch=True, on_step=True, logger=True)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)


def print_auto_logged_info(r):
    print("metrics: {}".format(r.data.metrics))


def main() -> None:
    mlflow.set_experiment("Autolog issue")

    # Initialize our model
    mnist_model = MNISTModel()

    # Initialize DataLoader from MNIST Dataset
    train_ds = MNIST(os.getcwd(), train=True,
                     download=True, transform=transforms.ToTensor())
    train_loader = DataLoader(train_ds, batch_size=32)

    # Initialize a trainer
    trainer = pl.Trainer(max_epochs=20, progress_bar_refresh_rate=20)

    # Auto log all MLflow entities
    mlflow.pytorch.autolog()

    # Train the model
    with mlflow.start_run() as run:
        trainer.fit(mnist_model, train_loader)

        # fetch the auto logged parameters and metrics
        print_auto_logged_info(mlflow.get_run(run_id=run.info.run_id))


if __name__ == "__main__":
    main()
```

### Other info / logs
PyTorch Lightning adds the `_step` and `_epoch` suffixes to the logged metrics, so on our side we used this info and slightly adjusted [__MLflowPLCallback](https://github.com/mlflow/mlflow/blob/86a03e9488f7cf2c7c0e46531fd2168923c79ec2/mlflow/pytorch/_pytorch_autolog.py#L83) by adding the following hook:
```
        def on_train_batch_end(
            self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx
        ):
           # ADD CODE TO FILTER ON `_step` SUFFIX

           # Then we log filtered metrics on pl_module.global_step instead of  pl_module.current_epoch
           metrics_logger.record_metrics(step_metrics, pl_module.global_step)
```
We also adjusted `__MLflowPLCallback._log_epoch_metrics()` to filter out metrics with the `_step` suffix.

Are there any plans to make classes such as [__MLflowPLCallback](https://github.com/mlflow/mlflow/blob/86a03e9488f7cf2c7c0e46531fd2168923c79ec2/mlflow/pytorch/_pytorch_autolog.py#L83) accessible so that we can subclass them to add/modify such functionality?

Thanks!

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by adamreeve**
## What changes are proposed in this pull request?

Add a new `log_every_n_step` parameter to `mlflow.pytorch.autolog` that enables logging step based metrics using the step count instead of epoch. By default, this parameter is `False` and these metrics will be logged using epoch numbers to keep backwards compatibility.

This fixes #4235.

This is a second attempt after #5497 was reverted as that PR accessed `trainer.logger_connector` which is being made private in the next PyTorch-Lightning version.

The new approach only uses `trainer.callback_metrics` for recent PyTorch-Lightning versions and additionally tracks which metrics have been logged on epochs so these aren't considered as step based metrics. Recording the metrics seen on epochs and steps is done outside the check for whether they should be logged in mlflow, so that eg. if `log_every_n_epoch` is greater than 1, we still know which metrics were logged on the epoch within PyTorch-Lightning and shouldn't be considered as step based metrics in the second epoch (and similarly for when steps are logged less frequently than epochs, but that seems quite unlikely).

I was a little concerned about the performance impact of retrieving and checking the metric keys on each step and epoch so did some benchmarking, and the impact was fairly negligible. I used a very simple PyTorch model with a single layer and random data and ran 100 epochs, with each epoch running 64 training batches. I'm logging to a local mlflow dev server with a sqlite tracking store. I also ran the same benchmarks after merging in the current changes from #5460 which enables batch SQL writing, and the extra time taken when logging on steps is greatly reduced:

| Test case | Time taken (s) | Time taken with batch writing (s) |
| ------------- | ------------- | ------------- |
| Logging epochs only | 26.6 | 18.8 |
| Logging every 10th step with #5497 | 62.7 | 19.0 |
| Logging every 10th step with this PR | 63.1 | 19.2 |

I had hoped that changing to this approach would mean that the code would be the same for all PyTorch-Lightning versions, but unfortunately versions prior to 1.4.0 had test failures and I found that `trainer.callback_metrics` accessed within a callback after a training batch didn't include the metrics from the current batch, as these are only updated after all callbacks have been run, so I had to revert back to using the `logger_connector` approach for older versions.

## How is this patch tested?

New unit test added.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

`mlflow.pytorch.autolog` now accepts a `log_every_n_step` parameter that can be used to log step based metrics against the step number.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by harupy — APPROVED** [review]
lgtm!
